File: app/src/main/java/com/example/dailycheck/util/ProgressUtils.kt
package com.example.dailycheck.util

import com.example.dailycheck.data.model.TaskEntity
import java.time.format.TextStyle
import java.util.Locale

data class DayStats(
    val total: Int,
    val completed: Int,
    val remaining: Int,
    val pct: Int? // null when total == 0
)

data class WeekSummary(
    val avgPct: Int?,
    val completed: Int,
    val created: Int,
    val bestDay: String?
)

data class Streaks(
    val current: Int,
    val best: Int
)

data class DayProgress(
    val dateStr: String,
    val dayName: String,
    val total: Int,
    val completed: Int,
    val pct: Int?
)

object ProgressUtils {
    fun computeDayStats(tasks: List<TaskEntity>): DayStats {
        val total = tasks.size
        val completed = tasks.count { it.completed }
        val pct = if (total == 0) null else ((completed.toFloat() / total.toFloat()) * 100).toInt()
        return DayStats(total = total, completed = completed, remaining = total - completed, pct = pct)
    }

    fun formatPct(pct: Int?): String = if (pct == null) "–" else "$pct%"

    fun motivationalMsg(stats: DayStats): String {
        if (stats.total == 0) return "Nothing planned yet."
        val pct = stats.pct ?: 0
        return when {
            pct >= 100 -> "Perfect day! 🎉"
            pct >= 80 -> "Excellent work! 💪"
            pct >= 60 -> "Good progress. Keep going."
            pct >= 40 -> "Keep moving forward."
            else -> "Every completed task counts."
        }
    }

    fun computeWeekSummary(allTasks: List<TaskEntity>, weekDates: List<String>): Pair<WeekSummary, List<DayProgress>> {
        val tasksByDate = allTasks.groupBy { it.date }
        var totalPctSum = 0
        var countedDays = 0
        var totalCompleted = 0
        var totalCreated = 0
        var bestDay: String? = null
        var bestPct = -1
        var bestCompleted = -1

        val dayProgressList = weekDates.map { dstr ->
            val date = DateUtils.parseDate(dstr)
            val dayName = date.dayOfWeek.getDisplayName(TextStyle.SHORT, Locale.getDefault())
            val dayTasks = tasksByDate[dstr] ?: emptyList()
            val st = computeDayStats(dayTasks)

            totalCompleted += st.completed
            totalCreated += st.total

            if (st.total > 0 && st.pct != null) {
                totalPctSum += st.pct
                countedDays++
                val better = st.pct > bestPct || (st.pct == bestPct && st.completed > bestCompleted)
                if (better) {
                    bestPct = st.pct
                    bestCompleted = st.completed
                    bestDay = date.dayOfWeek.getDisplayName(TextStyle.FULL, Locale.getDefault())
                }
            }

            DayProgress(
                dateStr = dstr,
                dayName = dayName,
                total = st.total,
                completed = st.completed,
                pct = st.pct
            )
        }

        val summary = WeekSummary(
            avgPct = if (countedDays > 0) Math.round(totalPctSum.toFloat() / countedDays) else null,
            completed = totalCompleted,
            created = totalCreated,
            bestDay = bestDay
        )

        return Pair(summary, dayProgressList)
    }

    fun computeStreaks(allTasks: List<TaskEntity>, todayStr: String = DateUtils.todayStr()): Streaks {
        val tasksByDate = allTasks.groupBy { it.date }
        val activeDates = tasksByDate.keys.filter { (tasksByDate[it]?.size ?: 0) > 0 }.sorted()

        if (activeDates.isEmpty()) return Streaks(0, 0)

        val minDate = activeDates.first()
        val maxDate = todayStr
        val threshold = 80

        // Current streak: walk backward from today, skipping zero-task days
        var current = 0
        var cursor = todayStr
        var guard = 0
        while (cursor >= minDate && guard < 20000) {
            val tasks = tasksByDate[cursor] ?: emptyList()
            val st = computeDayStats(tasks)
            if (st.total > 0) {
                if (st.pct != null && st.pct >= threshold) {
                    current++
                    cursor = DateUtils.addDays(cursor, -1)
                } else {
                    break
                }
            } else {
                cursor = DateUtils.addDays(cursor, -1)
            }
            guard++
        }

        // Best streak: scan chronologically
        var best = 0
        var running = 0
        var cur = minDate
        guard = 0
        while (cur <= maxDate && guard < 20000) {
            val tasks = tasksByDate[cur] ?: emptyList()
            val st = computeDayStats(tasks)
            if (st.total > 0) {
                if (st.pct != null && st.pct >= threshold) {
                    running++
                    if (running > best) best = running
                } else {
                    running = 0
                }
            }
            cur = DateUtils.addDays(cur, 1)
            guard++
        }

        if (current > best) best = current
        return Streaks(current = current, best = best)
    }
}

File: src/utils/progressUtils.ts
import type { DayData, DayStats, Task } from "../types";
import { weekdayFull } from "./dateUtils";
import { resolveDayData } from "./recurrenceUtils";

/**
 * Zero-task rule: a day with no tasks has pct = null. Callers must never
 * treat null as 0% or 100% — it means "no tasks", not "no progress" or
 * "fully done".
 */
export function dayStats(day: DayData | undefined): DayStats {
  const tasks = day?.tasks ?? [];
  const total = tasks.length;
  const completed = tasks.filter((t) => t.completed).length;
  const pct = total === 0 ? null : Math.round((completed / total) * 100);
  return { total, completed, remaining: total - completed, pct };
}

export function formatPct(pct: number | null): string {
  return pct === null ? "\u2013" : pct + "%";
}

export function motivationalMsg(stats: DayStats): string {
  if (stats.total === 0) return "Nothing planned yet.";
  const pct = stats.pct as number;
  if (pct >= 100) return "Perfect day! \uD83C\uDF89";
  if (pct >= 80) return "Excellent work! \uD83D\uDCAA";
  if (pct >= 60) return "Good progress. Keep going.";
  if (pct >= 40) return "Keep moving forward.";
  return "Every completed task counts.";
}

export interface WeekSummary {
  /** null when no day in the week has any tasks. */
  avgPct: number | null;
  completed: number;
  created: number;
  bestDay: string | null;
}

/**
 * Weekly average only counts days that have at least one task. A day with
 * zero tasks contributes to neither the average nor "best day", and can
 * never itself be the best day.
 */
export function weekSummary(
  days: Record<string, DayData>,
  weekDates: string[],
  recurringTasks?: Task[]
): WeekSummary {
  let totalPct = 0;
  let countedDays = 0;
  let completed = 0;
  let created = 0;
  let bestDay: string | null = null;
  let bestPct = -1;
  let bestCompleted = -1;

  weekDates.forEach((dstr) => {
    const day = recurringTasks ? resolveDayData(dstr, days[dstr], recurringTasks) : days[dstr];
    const st = dayStats(day);
    completed += st.completed;
    created += st.total;
    if (st.total === 0 || st.pct === null) return; // zero-task day: excluded entirely

    totalPct += st.pct;
    countedDays++;

    if (st.pct > 0) {
      const better = st.pct > bestPct || (st.pct === bestPct && st.completed > bestCompleted);
      if (better) {
        bestPct = st.pct;
        bestCompleted = st.completed;
        bestDay = weekdayFull(dstr);
      }
    }
  });

  return {
    avgPct: countedDays ? Math.round(totalPct / countedDays) : null,
    completed,
    created,
    bestDay
  };
}
